### PR TITLE
 [11.x] Allow adding array or string for web and api routes in bootstrap/app.php

### PR DESCRIPTION
**Pull request by mrthito**
> This PR is Resubmitted after Taylor feedback

This pull request proposes an enhancement to Laravel's `bootstrap/app.php` file, enabling developers to add routes in array format or as strings for both web and api route groups.

### Motivation

The motivation behind this enhancement is to provide developers with greater flexibility and convenience when defining routes, particularly in scenarios where applications have different guards and need to create different routes for each guard rather than writing require/require_once in route file.

### Use Cases

In projects where route management becomes complex due to the proliferation of routes or diverse authentication scenarios, this will simplifies the process by enabling developers to organize routes into multiple route files effortlessly. By allowing routes to be defined in array format, developers can easily create separate route files corresponding to different functionalities, user roles, or authentication guards. This approach promotes code organization and maintainability, as each route file can focus on a specific aspect of the application, making it easier to understand and manage.

## Additional Context

I'm developing an ecommerce website with five different guards, and I want to include separate route files for each guard. Using `require` or `require_once` made my code look outdated, and I disliked it. Let's simplify by allowing arrays as well in the `$web` parameter of `withRouting()` method in `Illuminate\Foundation\Application`

carbon

### CODE REVIEW

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php [hunk 1]
```diff
@@ -141,16 +141,16 @@ public function withBroadcasting(string $channels, array $attributes = [])
      * @return $this
      */
     public function withRouting(?Closure $using = null,
-        ?string $web = null,
-        ?string $api = null,
+        string|array $web = null,
```
Comment (PerryvanderMeer) [inline src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:144]: We still need `|null` here.

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php [hunk 2]
```diff
@@ -141,16 +141,16 @@ public function withBroadcasting(string $channels, array $attributes = [])
...
+        string|array $api = null,
```
Comment (PerryvanderMeer) [inline src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:145]: We still need `|null` here.

File: src/Illuminate/Foundation/Configuration/ApplicationBuilder.php [hunk 3]
```diff
@@ -174,24 +174,32 @@ public function withRouting(?Closure $using = null,
...
      * @param  string|null  $pages
      * @param  string|null  $health
      * @param  string  $apiPrefix
      * @param  callable|null  $then
      * @return \Closure
      */
-    protected function buildRoutingCallback(?string $web,
-        ?string $api,
+    protected function buildRoutingCallback(string|array $web = null,
+        string|array $api = null,
```
Comment (PerryvanderMeer) [inline src/Illuminate/Foundation/Configuration/ApplicationBuilder.php:186]: We still need `|null` here.

**Comment by driesvints** [comment]
We'll have to revert this one unfortunately because of https://github.com/laravel/framework/issues/51406

**Pull request merged**